File: backend/src/main/java/com/ll/hotel/global/request/Rq.java
package com.ll.hotel.global.request;

import com.ll.hotel.domain.member.member.entity.Member;
import com.ll.hotel.domain.member.member.repository.MemberRepository;
import com.ll.hotel.global.exceptions.ErrorCode;
import com.ll.hotel.global.security.oauth2.dto.SecurityUser;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.context.annotation.RequestScope;

import java.util.Optional;

@Component
@RequestScope
@RequiredArgsConstructor
public class Rq {
    private final MemberRepository memberRepository;
    private Member actor;

    public Member getActor() {
        if (actor == null) {
            actor = Optional.ofNullable(SecurityContextHolder.getContext().getAuthentication())
                    .map(Authentication::getPrincipal)
                    .filter(principal -> principal instanceof SecurityUser)
                    .map(principal -> (SecurityUser) principal)
                    .flatMap(securityUser -> memberRepository.findByMemberEmail(securityUser.getEmail()))
                    .orElseThrow(ErrorCode.UNAUTHORIZED::throwServiceException);
        }

        return actor;
    }
}

File: backend/src/main/java/com/ll/hotel/global/request/Rq.kt
package com.ll.hotel.global.request

import com.ll.hotel.domain.member.member.entity.Member
import com.ll.hotel.domain.member.member.repository.MemberRepository
import com.ll.hotel.global.exceptions.ErrorCode
import com.ll.hotel.global.security.oauth2.dto.SecurityUser
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.context.annotation.RequestScope

@Component
@RequestScope
class Rq(
    private val memberRepository: MemberRepository
) {
    private var actor: Member? = null

    fun getActor(): Member {
        if (actor == null) {
            actor = SecurityContextHolder.getContext().authentication
                ?.takeIf { it.principal is SecurityUser }
                ?.principal
                ?.let { it as SecurityUser }
                ?.let { memberRepository.findByMemberEmail(it.email).orElse(null) }
                ?: throw ErrorCode.UNAUTHORIZED.throwServiceException()
        }
        return actor!!
    }
}